ext/src/http/client/curl/http_client_curl.cc

```diff
@@ -57,11 +60,85 @@ nostd::shared_ptr<HttpCurlGlobalInitializer> HttpCurlGlobalInitializer::GetInsta
   return shared_initializer;
 }
 
+#ifdef ENABLE_OTLP_COMPRESSION_PREVIEW
+// Original source:
+// https://stackoverflow.com/questions/12398377/is-it-possible-to-have-zlib-read-from-and-write-to-the-same-memory-buffer/12412863#12412863
+int deflateInPlace(z_stream *strm, unsigned char *buf, uint32_t len, uint32_t *max_len)
+{
+  // must be large enough to hold zlib or gzip header (if any) and one more byte -- 11 works for the
+  // worst case here, but if gzip encoding is used and a deflateSetHeader() call is inserted in this
+  // code after the deflateReset(), then the 11 needs to be increased to accommodate the resulting
+  // gzip header size plus one
+  std::array<unsigned char, 11> temp{};
+
```

Comment:
The original code from Mark Adler has:

```
    /* initialize deflate stream and point to the input data */
    ret = deflateReset(strm);
    if (ret != Z_OK)
        return ret;
```

was it forgotten, or is this not needed and why ?

Reply:
From the [zlib manual](https://www.zlib.net/manual.html#Advanced):

> ZEXTERN int ZEXPORT deflateReset(z_streamp strm);
>
> This function is equivalent to deflateEnd followed by deflateInit, but does not free and reallocate the internal compression state. The stream will leave the compression level and any other attributes that may have been set unchanged. total_in, total_out, adler, and msg are initialized.
>
> deflateReset returns Z_OK if success, or Z_STREAM_ERROR if the source stream state was inconsistent (such as zalloc or state being NULL).

Thus, the use of `deflateReset()` does not fit the use case well here because it implies calling `deflateInit()`.
Since we additionally need to supply a _windowBits_ argument, which is only supported in `deflateInit2()`, the setup was kept "out" of this in-place deflate method as originally implemented in `SendRequest()` on [L165](https://github.com/open-telemetry/opentelemetry-cpp/pull/3178/files/fadff90c126777a4c68480c8c725847f2963d7a6#diff-21cfd22f6082a873195bac59990cb9f311c98451cba9f2ac3cb48555c5f960eaR165) followed by `deflateEnd()` on [L191](https://github.com/open-telemetry/opentelemetry-cpp/pull/3178/files/fadff90c126777a4c68480c8c725847f2963d7a6#diff-21cfd22f6082a873195bac59990cb9f311c98451cba9f2ac3cb48555c5f960eaR191), once the compression is complete.